git_theta/lsh/e2lsh.py

```diff
@@ -0,0 +1,89 @@
+"""Classes for computing Euclidean locality-sensitive hashes"""
+
+import numpy as np
+import numba as nb
+import os
+
+from git_theta.lsh import HashFamily, RandomnessPool
+from git_theta.lsh.types import Signature, Parameter
+from git_theta.utils import EnvVarConstants
+
+
+class E2LSH(HashFamily):
+    """
+    Class for performing the Euclidean l2 LSH (E2LSH) algorithm described in https://www.cs.princeton.edu/courses/archive/spring05/cos598E/bib/p253-datar.pdf
+    with a pre-computed randomness pool as described in Section 3 of http://personal.denison.edu/~lalla/papers/online-lsh.pdf
+
+    Also see http://mlwiki.org/index.php/Euclidean_LSH for an introduction to E2LSH
+    """
+
+    def __init__(self, signature_size: int, bucket_width: float):
+        super().__init__(signature_size)
+        self.bucket_width = bucket_width
+
+    @property
+    def name(self) -> str:
+        return "E2LSH"
+
+    def hash(self, x: Parameter) -> Signature:
+        """Convert `x` to its signature."""
+        x = x.ravel()
+        hyperplanes = self.pool.get_hyperplanes(x.size)
+        return np.rint((x @ hyperplanes) / self.bucket_width).astype(np.int64).tolist()
+
+    def distance(self, query: Signature, data: Signature) -> float:
+        """Compute the distance between two E2LSH signatures"""
+        query, data = np.array(query), np.array(data)
+        return (
+            (1 / np.sqrt(self.signature_size))
+            * np.linalg.norm(query - data)
+            * self.bucket_width
+        )
+
+
+class FastE2LSH(E2LSH):
+    """
+    Class for performing the Euclidean LSH using numba-jitted loops. This is both faster than the E2LSH class using numpy matrix multiplications and
+    also uses less memory since the whole hyperplane matrix (feature_size x signature_size) is never in memory at once.
+    """
+
+    def name(self) -> str:
+        return "FastE2LSH"
+
+    def hash(self, x: Parameter) -> Signature:
+        """Convert `x` to its signature."""
+        return nb_hash(
+            x.ravel(), self.signature_size, self.pool, self.bucket_width
+        ).tolist()
+
+
+@nb.jit(nopython=True, parallel=True)
+def nb_hash(
+    x: Parameter, signature_size: int, pool: RandomnessPool, bucket_width: float
+) -> Signature:
+    signature = np.zeros(signature_size)
+
+    for signature_idx in nb.prange(signature_size):
+        for feature_idx in range(x.size):
+            hyperplane_element = pool.get_hyperplane_element(feature_idx, signature_idx)
+            signature[signature_idx] += x[feature_idx] * hyperplane_element
+
+    return np.rint(signature / bucket_width).astype(np.int64)
+
+
+def get_signature_size():
+    signature_size = os.environ.get(EnvVarConstants.LSH_SIGNATURE_SIZE)
+    return int(signature_size) if signature_size else 16
```

Comment:
From testing it seems like approximating the L_2 distance with just 16 random projections is pretty accurate. We also want to keep this as small as possible because we don't want to do too many projections of larger parameter groups (e.g., an embedding matrix of size 50000x1000 ~ 50M parameters). Running the non-numba LSH implementation on tensors of this size is extremely slow (at least on my laptop). With the numba implementation it takes about half a second to hash this tensor which isn't great but is usable.